File: house_robber/houseRobber.py
"""
Python implementation of top-down house robber
"""
class houseRobber:
    def rob(self, nums: List[int]) -> int:
        self.memo_list = [-1] * 100
        return self.helper(nums, 0);
    
    def helper(self, nums: List[int], pos) -> int:
        if (pos >= len(nums)):
            return 0
        if (self.memo_list[pos] != -1):
            return self.memo_list[pos]
        pick = nums[pos] + self.helper(nums, pos + 2)
        skip = self.helper(nums, pos + 1)
        if (pick > skip):
            self.memo_list[pos] = pick
            return pick
        else:
            self.memo_list[pos] = skip
            return skip

File: house_robber/houseRobber.java
/**
 * HouseRobber - given an array of house values, get the max value from robbing.  Cannot rob houses
 * that are adjacent to each other.
 *
 * Top down approach, using memoization.  Array default set to -1 as 0 is a possible reward.
 * Base case: if reach the end of the array, return 0.
 * At each step, decide to choose to rob this house (pick) or not (skip).
 * */
class HouseRobber {
    int[] arr = new int[100];
    public int rob(int[] nums) {
        Arrays.fill(arr, -1);
        return helper(nums, 0);
    }
    public int helper(int[] nums, int pos) {
        if (pos >= nums.length) {
            return 0;
        }
        if (arr[pos] != -1) {
            return arr[pos];
        }
        int pick = nums[pos] + helper(nums, pos + 2);
        int skip = helper(nums, pos + 1);
        if (pick > skip) {
            arr[pos] = pick;
            return pick;
        } else {
            arr[pos] = skip;
            return skip;
        }
    }
}
